BUG: optimize: add py314 wrapper for `inspect.signature` of pub callables

willwiz opened the pull request:
#### Reference issue
Closes gh-23913

#### What does this implement/fix?
Add a wrapper for `inspect.signature` with the keyword parameter `annotation_format=annotationlib.Format.FORWARDREF` if Python version >= 3.14. 

PEP 649/749, lazy evaluation of annotations, allows for scenarios where users may pass callables with annotations that are only defined when type checking or are forward-referenced. `inspect.signature` will raise a NameError in these cases, but scipy does need the annotation information.  Python 3.14 also adds the keyword parameter `annotation_format` to `inspect.signature`. Here, `annotation_format=annotationlib.Format.FORWARDREF` gives the closest behavior to pre-Python 3.14, and will return a `ForwardRef` object. None of the functions affected by this change needs the annotation information, so this should be sufficient to preserve existing functionality. Functions that are internal or for testing should not be affected. 

#### Additional information

Changes:
scipy/optimize/_minimize.py:minimize:741
- inspects user-defined `callback` parameter

scipy/optimize/_optimize.py:_wrap_callback:93
- wraps the `callback` parameter for most algorithms

scipy/optimize/_cobyla_py.py:_minimize_cobyla:249
- skips _wrapped_callback and inspects naked `callback` parameter

scipy/optimize/_basinhopping.py:BasinHoppingRunner._monte_carlo_step:132
- inspects `accept_test` keyword argument from scipy.optimize.basinhopping
- Notes: It checks if the signature matches the exact pattern of `Metropolis`. As long as `inspect.signature` doesn't raise, the actual signature of `accept_test` arguments does not matter, and should match previous functionality. 

{scipy/optimize/_minpack_py.py:curve_fit:904,scipy/optimize/_optimize.py:brute:3674}
- Uses `getfullargspec_no_self` but doesn't need its full functionality. Minor changes to match previous functionality. 

External modules:
The cobyqa submodule (scipy/_lib/cobyqa) also has this issue and skips `_wrap_callback` when called from minimize. 

Side notes on other instances where this can happen in scipy: 

scipy/stats/_resampling.py:_bootstrap_iv:176
- takes callable parameter `statistic` in scipy.stats.bootstrap
scipy/stats/_resampling.py:_monte_carlo_test_iv:724
- takes callable parameter `statistic` in scipy.stats.monte_carlo_test
scipy/stats/_resampling.py:{_wrap_kwargs:1032, _power_iv:1097}
- takes callable parameters (`test`, `rvs`) in scipy.stats.power
scipy/stats/_resampling.py:_permutation_test_iv:1032
- takes callable parameters `statistic` in scipy.stats.permutation_test
scipy/stats/_sensitivity_analysis.py:sobol_indices:618
- keyword `method` can be a user-defined callable.

willwiz committed:
BUG: optimize: add py314 wrapper for inspect.signature of pub callables.

wrapper addresses lazy evaluation of annotations (PEP 649/749) by
adding new keyword argument annotation_format for forward refs in
Python 3.14 or greater. See gh-23913.

willwiz committed:
MAINT:optimize: fix lint issues for imports

willwiz committed:
BUG: optimize: curve_fit: inspect behavior did not match previous

The last edit for improving the compatibility of inspect.signature did
not account of the fact that curve_fit only consider some parameter
types, allowing for inconsistent behavior in some cases, e.g. **kwargs

nickodell commented on scipy/_lib/_util.py:
```diff
@@ -60,6 +60,28 @@
+else:
+    def wrapped_inspect_signature(callable):
+        """Get a signature object for the passed callable."""
+        return inspect.signature(callable)
```
```suggestion
    wrapped_inspect_signature = inspect.signature
```

nickodell reviewed (changes requested):
Nice work here. Can you add a test for this behavior?

I would like to see a test which fails on 3.14 if this patch were not applied.

The file `scipy/scipy/optimize/tests/test_optimize.py` is probably a sensible place for it.

Since the test is testing something which is a NameError in a previous version of Python, you might need to write the test such that it selectively defines an import if running on pre-Python 3.14 systems.

e.g.

```
if sys.version < (3, 14):
    from typing import List

def callback(...):  # Uses List in an annotation somehow
    ...
```

This is just an example, you can probably think of a better idea. (Perhaps you could also selectively define the test in an if?)

willwiz committed:
BUG: optimize: adjustment for inspect signature fix

willwiz committed:
TST: optimize: test for error due to undefined annotations of user callables

See gh-23913.
Tests adds wrapper with undefinite annotations to callables that can be passed
into optimize.minimize and other public functions, then check if a NameError
is triggered. Fixed in gh-23922.

willwiz commented:
The suggested changes are integrated. Also added the tests for these issues. 

These tests should be skipped in 3.11 -- or at least they are on my computer. There should be 17 fails on 3.14 without these changes -- `method="tnc"` currently does not implement this type of inspection.

nickodell commented:
> The suggested changes are integrated. Also added the tests for these issues.

Thanks. I see a few test failures that are caused by issues elsewhere. See #23932

>These tests should be skipped in 3.11 -- or at least they are on my computer. There should be 17 fails on 3.14 without these changes - method="tnc" currently does not implement this type of inspection.

Sorry - by "should" do you mean that they are being skipped right now, or that they ought to be skipped?

willwiz commented:
> > These tests should be skipped in 3.11 -- or at least they are on my computer. There should be 17 fails on 3.14 without these changes - method="tnc" currently does not implement this type of inspection.
> 
> Sorry - by "should" do you mean that they are being skipped right now, or that they ought to be skipped?

Sorry, I worded that a bit weirdly. 

I marked these tests to be skipped with pytest if the version is < 3.14 with this
```
@pytest.mark.skipif(sys.version_info < (3, 14), reason="Requires PEP 649/749 from Python 3.14+.")
```
However, I only locally tested that these changes behave correctly with spin on Python 3.11 and 3.14. I am assuming that this should be fine for other setups and the tests that are done on CI.

nickodell committed:
Merge remote-tracking branch 'upstream/main' into callbacksig

nickodell commented:
Hi - Sorry for going silent for a few days. I'm still reviewing this.

I have a few TODOs before merging.

- [ ] Check that removing annotations did not break docs (One user of getfullargspec_no_self is inside `docs/`.)
- [x] Check that these changes have statement / branch coverage

   - _lib/util.py:545
   - basinhopping.py:133
   - _cobyla_py.py:248
   - _minimize:741
   - _optimize.py:93

- [ ] Check that we're correct about the versions of Python where the syntax works and fails.
- [x] Check that attached test fails without the patch.
- [ ] This is a fairly subtle change with different behavior on different versions, so I'd like to get CI green before merging.
   - [ ] Pythan failure: https://github.com/scipy/scipy/actions/runs/19222125724/job/54941950948?pr=23922#step:7:344
   - [ ] Rtools failure: https://github.com/scipy/scipy/actions/runs/19222125699/job/54941947556?pr=23922#step:4:38
   
   I think that covers all the failures without existing issues. (Boy, this is really the week when all our tooling decided to quit on us.)

nickodell committed:
TST: optimize: avoid catching errors in tests

nickodell committed:
TST: optimize: run tests for fewer iterations

[skip ci]

nickodell committed:
Merge remote-tracking branch 'upstream/main' into callbacksig

willwiz commented:
In regards to the tach error with `annotationlib`, tach gets its list of std libs from the Python interpreter as 
```python
set(sys.builtin_module_names).update(sys.stdlib_module_names).update(KNOWN_MODULE_SPECIAL_CASES)
```
On my end, locally, tach doesn't flag `annotationlib` if I am using Python 3.14, but does flag it if I'm on Python 3.11. It seems like the logic for handling version-dependent conditional imports is not there. Maybe, it would be more appropriate to add `annotationlib` temporarily to tach.toml under `[external] exclude` until support for Python 3.13 is dropped?

nickodell commented:
> Maybe, it would be more appropriate to add `annotationlib` temporarily to tach.toml under `[external] exclude` until support for Python 3.13 is dropped?

That seems reasonable to me. Go for it.

But I think you and I have different definitions of temporary. :) We will be dropping support for 3.13 sometime in 2028.

willwiz committed:
DEV: tach: add `annotationlib` to external.exclude

This is necessary to prevent dtach from flagging import of `annotationlib` while maintaining
backward compatibility for Python <3.14.
First instance is in gh-23922.

nickodell reviewed (approved):
Thanks, looks good to me. As noted [here](https://github.com/scipy/scipy/pull/23922#issuecomment-3509634492), there is a test failure remaining, but it doesn't look very scary to me.

willwiz committed:
Merge branch 'main' into callbacksig

lucascolley committed:
Merge branch 'main' into callbacksig

willwiz committed:
Merge branch 'main' into callbacksig

willwiz committed:
TST/BUG: optimize.annotations: set seed for rng for basinhopping

test_basinhopping_annotations can sometime fail to converge within tolerance depending,
triggering convergence error (see gh-23922).

nickodell commented:
> I'm not sure about the errors with `pixi`, and

Probably [Cloudflare](https://downdetector.com/status/cloudflare/) related. prefix.pixi.dev is hosted on Cloudflare.

> I guess the build doc errors on https request to numpy are kind of just bad luck as well?

NumPy.org is hosted on GitHub pages and Cloudflare. So that could be the cause of this intersphinx error:

```
WARNING: failed to reach any of the inventories with the following issues:
intersphinx inventory 'https://numpy.org/neps/objects.inv' not fetchable due to <class 'requests.exceptions.HTTPError'>: 500 Server Error: Internal Server Error for url: https://numpy.org/neps/objects.inv
WARNING: failed to reach any of the inventories with the following issues:
intersphinx inventory 'https://numpy.org/doc/stable/objects.inv' not fetchable due to <class 'requests.exceptions.HTTPError'>: 500 Server Error: Internal Server Error for url: https://numpy.org/doc/stable/objects.inv
WARNING: failed to reach any of the inventories with the following issues:
intersphinx inventory 'https://matplotlib.org/stable/objects.inv' not fetchable due to <class 'requests.exceptions.HTTPError'>: 500 Server Error: Internal Server Error for url: https://matplotlib.org/stable/objects.inv
myst v4.0.1: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions={'dollarmath', 'l
```

If it can't get the NumPy inventory, then it will not be able to tell if any of the NumPy doc links are correct.

For the [GitHub checkout errors](https://downdetector.com/status/github/), I think something similar is going on there.

nickodell commented:
@willwiz I wouldn't worry about that one - I don't think it's plausibly downstream of this change.

@andyfaff I'm inclined to merge this via 'Squash and merge,' to get rid of some of the merge noise above. Do you want to take more time to review it, or are you happy with it as-is?

andyfaff commented:
I won't have time to look at it again.
